_source/_posts/2020-12-16-react-login.md

```diff
@@ -0,0 +1,599 @@
+tweets:
+- "Need a login form in your @reactjs app? This tutorial will hook you up!"
+- "Login is a common feature. Learn how to add login to your React apps with this quick guide."
+- "A quick guide to login options with @reactjs + @okta. 👇 💙"
+image: blog/react-login/react-login.png
+type: conversion
+---
+
+Almost any web app needs some sort of access control, usually implemented by user login. Choosing how user authentication is implemented depends on the type of application and its audience. In this post, I want to show you a few different ways of creating a login feature in a single-page React application using Okta.
+
+I will start with a login redirect. This is the easiest option to implement and is a good choice for some use cases. On the other hand, hosting the login form inside of the application can provide a more seamless user experience. The Okta Sign-In Widget can be used as a simple out-of-the-box login form. To have more control over the appearance of the user login, you can also implement your own form and use Okta's authentication service to verify the user's credentials.
+
+For this tutorial, I will assume that you are familiar with JavaScript and have some knowledge of React. You should have Node and `npm` installed on your system.
+
+**Prerequisites**:
+
+- [Node 14](https://nodejs.org/)
+- [Okta CLI 0.7.1+](https://github.com/okta/okta-cli)
+
+**Table of Contents**{: .hide }
+* Table of Contents
+{:toc}
+
+## Get a Free Okta Developer Account
+
+Before you get started, you need to sign up for a free Okta developer account. If you haven't already got one, go to [developer.okta.com](https://developer.okta.com/), click **Sign Up**, and follow the sign-up procedure. Once completed, you can log in to the Developer Console.
+
+**TIP**: You can also create an account and register apps using the [Okta CLI](https://github.com/okta/okta-cli#readme) with the `okta register` and `okta apps create` commands.
+
+In your Okta Development Console, select **Applications** from the top menu and click **Add Application**. Next, select **Single-Page App** and click **Next**. You now see a form with application settings. You can give the application any name of your choice. Both the **Base URI** and the **Logout redirect URI** should be changed to `http://localhost:3000/`.
+
+You will be testing the application using the React development server that listens on port 3000 on your local machine. For production, you would of course change this to the URI of the production server. The **Login redirect URI** should be `http://localhost:3000/callback`.
+
+Click **Done** when ready. On the next page, you will see an overview of the settings as well as a **Client ID**. Note the ID for later when you set up your React application.
+
+## Create a React Application for Login
+
+Now you are ready to create the React application. Open a terminal in a folder of your choice and run the following command.
+
+```bash
+npx create-react-app react-login
```

Comment:
Is this supposed to be `npm` or is there some new `npx` package that hasn't been mentioned yet?

Reply:
Ah! TIL: https://nodejs.dev/learn/the-npx-nodejs-package-runner/

Reply:
So. npx was new to me. Though it _has_ been out for 3 years. If it were up to me, I'd add a short sentence above saying something like "... run npx, the Node.js Package Runner included with npm"

Reply:
Will do.